src/solvers/dgsem_unstructured/sort_boundary_conditions.jl

```diff
@@ -41,17 +41,23 @@ function initialize!(boundary_types_container::UnstructuredSortedBoundaryTypes{N
 
   unique_names = unique(cache.boundaries.name)
 
-  # TODO: This needs to be handled differently for the `ParallelP4estMesh` since the boundaries
-  # are distributed and thus unique_names only contains the names of boundaries on the local process
-  # See https://github.com/trixi-framework/Trixi.jl/issues/1047
-  if !mpi_isparallel()
-    for key in keys(boundary_dictionary)
-      if !(key in unique_names)
-        error("Key $(repr(key)) is not a valid boundary name")
+  if mpi_isparallel()
+    # Exchange of boundaries names
+    for i in 0:(mpi_nranks()-1)
+      if i!=mpi_rank()
+        MPI.isend(unique_names, i, 0, mpi_comm())
+        recv_names = MPI.recv(mpi_comm(); source=i, tag=0)
+        unique_names = unique(vcat(unique_names, recv_names))
       end
     end
```

Comment:
Here you are manually sending the data to all other processes. I think this can be potentially improved in two ways:
* Instead of using many P2P communications, you can reduce this probably to a single call to [`MPI.Allgatherv!`](https://juliaparallel.org/MPI.jl/v0.20/reference/collective/#MPI.Allgatherv!). Usually this makes the code easier to read and the communication slightly faster (although the latter is not really a big issue here, probably)
* Instead of sending all data to every rank, then do the check there, and then potentially fail on each rank, you can also just gather all data on the root rank (i.e. on the process with rank zero) and then call an [`MPI.Abort`](https://juliaparallel.org/MPI.jl/v0.20/reference/environment/#MPI.Abort) from there.

Reply:
Thank you for review!
There are two reasons, why I haven't used the [MPI.Allgatherv!](https://juliaparallel.org/MPI.jl/v0.20/reerence/collective/#MPI.Allgatherv!).
1.  To collect data with `MPI.Allgatherv!` or `MPI.Gatherv!`, I have to explicitly specify the sizes of every data-chunk in receive_[MPI.VBuffer](https://juliaparallel.org/MPI.jl/v0.20/reerence/collective/#MPI.VBuffer). I think, it's possible only with additional communication via `MPI.Allgather` or `MPI.Gather`.
2.  All collective communications (Gather, Scatter, ...) in `MPI.jl` are working only with `isbitstype`. And sent `Vector{Symbol}` is non-bitstype, I have tried to serialize it with `MPI.serialize`, but even so I get the error `Type must be isbitstype`. `MPI.isend` and `MPI.recv` take care of serialization themselves. 

I will try to find some way to serialize `Vector{Symbol}`, but it also seems to be not easy read.

Reply:
1. One additional `MPI.Gather` would be OK. For large processor counts, this should still be cheaper than P2P  (O(log n) vs. O(n)).
2. Are you sure? Maybe you can ask what the "canonical" approach should be for an `MPI.Gatherv!` of `Symbol`s in the `#hpc` channel on the Julia slack and `cc @sloede` me

Reply:
1. I implemented a Dissemination pattern, that also should be O(log(n)):
```
using MPI
MPI.Init()
comm=MPI.COMM_WORLD
rank=MPI.Comm_rank(comm)
size=MPI.Comm_size(comm)
for i in 0:ceil(Int,log(size))
  dest = (rank+2^(i))%size
  src =  (rank-2^(i)+size)%size
  MPI.isend(unique_names, dest, 0, comm)
  recv_names = MPI.recv(comm; source=src, tag=0)
  unique_names = unique(vcat(unique_names, recv_names))
end
```
2. I asked this question in `#hpc`. I'm waiting for an answer.

Reply:
Yes, but all this code is still a lot more verbose than a single `MPI.Gatherv!` :-) Mosè suggested to use a vector of strings, have you tried that? I.e., by converting them with something like
```julia
names_as_symbols = Symbol[:a, :b, :c]
names_as_strings = string.(names)
```

Reply:
Yes, I did.
I wrote the simple test:
```
using MPI
MPI.Init()
names_as_symbols = Symbol[:a, :b, :c]
names_as_strings = string.(names)
comm=MPI.COMM_WORLD
if MPI.Comm_rank(comm) == 0
    buf=Vector{String}(undef, 3*3)
    MPI.Gatherv!(MPI.IN_PLACE, MPI.VBuffer(buf, [3,3,3]), comm; root=0)
else
    buf_symb=Symbol[:a, :b, :c]
    buf_str=string.(buf_symb)
    MPI.Gatherv!(buf_str, nothing, comm; root=0)
end
```
It crashes with `LoadError: ArgumentError: Type must be isbitstype`

Reply:
From the Julia Slack (reposted here to conserve):
```julia
julia> v_as_symbols = [:a, :b, :c]
3-element Vector{Symbol}:
 :a
 :b
 :c

julia> v_as_bytes = Vector{UInt8}(join(v_as_symbols, "\0"))
5-element Vector{UInt8}:
 0x61
 0x00
 0x62
 0x00
 0x63

julia> # send as bytes via MPI

julia> back_as_symbols = Symbol.(split(String(v_as_bytes), "\0"))
3-element Vector{Symbol}:
 :a
 :b
 :c
```
Maybe not the prettiest approach in terms of allocations, but might work.

Reply:
Thank you very much! I saw it and already implemented:
    ```
send_byte=Vector{UInt8}(join(unique_names, "\0"))
    push!(send_byte,0)
    if mpi_rank() == 0
      buf_size=MPI.Gather(length(send_byte), mpi_comm(); root=0)
      buf=Vector{UInt8}(undef, sum(buf_size))
      MPI.Gatherv!(send_byte, MPI.VBuffer(buf, buf_size), mpi_comm(); root=0)
      new_names=Symbol.(split(String(buf), "\0"))
      pop!(new_names)
      unique_names=unique(new_names)
    else
      MPI.Gather(length(send_byte), mpi_comm(); root=0)
      MPI.Gatherv!(send_byte, nothing, mpi_comm(); root=0)
    end
```
There is one additional not pretty correction, that we have to add our split-symbol in the end of every `unique_names`, otherwise last element of previous vector and first one of the next vector cannot be splitted.

Reply:
Ah, maybe you can just `append!(send_byte, 0)` or something similar...

Reply:
Yes, I used `push!`